src/docs/binding-generics.md

```diff
@@ -0,0 +1,189 @@
+# Binding generics
+
+This document explores projections of Swift generics into C#.
+
+1. [Generic functions](#generic-functions)
+2. Generic types (TBD)
+3. Generics with protocol constraints (TBD)
+4. Generics with PAT constraints (TBD)
+
+## Generic functions
+
+When a generic type is used in Swift, its memory layout isn't known at compile time. To manage this, the method parameter whose type is generic represented by an opaque pointer. To handle this:
+
+- **Opaque Pointers**: Swift represents generic parameters as opaque pointers (`%swift.opaque`), abstracting the actual memory layout.
+- **Type Metadata**: Swift includes a type metadata pointer (`%T`) as an implicit argument at the end of a function's signature. This metadata provides the runtime information needed to manage the generic type.
+- **Indirect Return**: For returning generic types, Swift uses an [indirect return result](https://github.com/dotnet/runtime/issues/100543), where the caller provides a memory location for the return value.
+
+Consider the following simple swift function:
+
+```swift
+func returnData<T>(data: T) -> T {
+    return data
+}
+```
+
+When compiled into LLVM-IR the above function has the following LLVM-IR signature:
+
+```llvm
+@"output.returnData<A>(data: A) -> A"(ptr noalias nocapture sret(%swift.opaque) %0, ptr noalias nocapture %1, ptr %T)
+```
+
+- `%0`: Indicates an **indirect return result**, where the caller allocates memory for the return value.
+- `%1`: The `data` parameter, passed as a pointer.
+- `%T`: The **type metadata** for `T`.
+
+### Projecting into C\#
+
+A sample projection into C# assuming that the type argument is projected as struct could look like this:
+
+```csharp
+[DllImport(Path, EntryPoint = "...")]
+[UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+public static extern void PInvoke_ReturnData(SwiftIndirectResult result, void* data, void* metadata);
+
+public static unsafe T ReturnData<T>(T data) where T : unmanaged, ISwiftObject
+{
+    var metadata = Runtime.GetMetadata(data);
+    nuint payloadSize = /* Extract size from metadata */;
+    
+    
+    var payload = NativeMemory.Alloc(payloadSize);
+    try 
+    {
+        var result = new SwiftIndirectResult(payload);
+        PInvoke_ReturnData(result, &data, metadata);
+        return *(T*)payload;
+    }
+    finally
+    {
+        NativeMemory.Free(payload);
+    }
+}
+```
+
+However, the projection has to support type arguments agnostic of whether they are projected as structs (frozen structs) or classes (non-frozen structs). We cannot then use the `unmanaged` constraint on `T` and will need to introduce abstractions for getting a native handle of an arbitrary Swift type and constructing a Swift type from such a handle.
+
+Those could look something like
+
+```csharp
+public static unsafe void* GetNativeHandle<T>(T type) where T : ISwiftObject
+public static unsafe T ConstructFromNativeHandle<T>(void* payload) where T : ISwiftObject
```

Comment:
For heap allocated types (classes and actors), the pointer is not a buffer but is is and actual instance pointer. For getting that into C#, we need to have a global registry that has a map of NativeHandle -> GCHandle where the GCHandle is taken from the actual C# instance of the type. For a type that doesn't exist in the map, we use the C# type (if we have it) or use the TypeMetadata to identify the C# type and run a flavor of constructor that takes a NativeHandle to initialize the type. As part of the memory management, we take both a strong and weak reference to the Swift handle. If the C# type is disposed, we inform the registry and it removes it from the cache as well as dropping the references. In BTfS this is in the class [SwiftObjectRegistry](https://github.com/xamarin/binding-tools-for-swift/blob/main/SwiftRuntimeLibrary/SwiftObjectRegistry.cs). The strong reference is taken in the object and the weak is taken in the registry.

Reply:
Thanks for the comments!

I don't really understand the `InitializeWithCopy` use case. In this scenario a new non-blittable struct is created on swift side and it being written to a buffer we provide from the C# side. So I assumed that just assigning this buffer to the C# projection should work? I cant fit where we would use `InitializeWithCopy` in this scenario. Could you please share some more context?

I would like to scope out heap allocated types from this from now and focus on structs as those represent a big chunk of our current needs. Then we would get back when we go and project classes / actors. However, I do think that once we start this description will be very relevant so I will go and include this comment it in the doc. What do you think?

Reply:
The InitializeWithCopy should be used if we need to return a type to a swift caller, for example a closure, virtual method - anything done from a reverse invoke.

Absolutely include this in the doc. I understand the focus on value types since that's most of what Apple prefers in Swift, but since the generic programming can take any type, we need to ensure that we build with an eye towards that.